chromadb/server/fastapi/__init__.py

```diff
@@ -205,153 +334,831 @@ def __init__(self, settings: Settings):
+        self._system.stop()
+
+    def app(self) -> fastapi.FastAPI:
+        return self._app
+
+    async def rate_limit_exception_handler(
+        self, request: Request, exc: RateLimitError
+    ) -> JSONResponse:
+        return JSONResponse(
+            status_code=429,
+            content={"message": "Rate limit exceeded."},
+        )
+
+    def root(self) -> Dict[str, int]:
+        return {"nanosecond heartbeat": self._api.heartbeat()}
+
+    async def quota_exception_handler(
+        self, request: Request, exc: QuotaError
+    ) -> JSONResponse:
+        return JSONResponse(
+            status_code=429,
+            content={
+                "message": f"quota error. resource: {exc.resource} "
+                f"quota: {exc.quota} actual: {exc.actual}"
+            },
+        )
+
+    async def heartbeat(self) -> Dict[str, int]:
+        return self.root()
+
+    async def version(self) -> str:
+        return self._api.get_version()
+
+    @trace_method(
+        "auth_request",
+        OpenTelemetryGranularity.OPERATION,
+    )
+    def auth_request(
+        self,
+        headers: Headers,
+        action: AuthzAction,
+        tenant: Optional[str],
+        database: Optional[str],
+        collection: Optional[str],
+    ) -> None:
+        """
+        Authenticates and authorizes the request based on the given headers
+        and other parameters. If the request cannot be authenticated or cannot
+        be authorized (with the configured providers), raises an HTTP 401.
+        """
+        if not self.authn_provider:
+            add_attributes_to_current_span(
+                {
+                    "tenant": tenant,
+                    "database": database,
+                    "collection": collection,
+                }
+            )
+            return
+
+        user_identity = self.authn_provider.authenticate_or_raise(dict(headers))
+
+        if not self.authz_provider:
+            return
+
+        authz_resource = AuthzResource(
+            tenant=tenant,
+            database=database,
+            collection=collection,
+        )
+
+        self.authz_provider.authorize_or_raise(user_identity, action, authz_resource)
+        add_attributes_to_current_span(
+            {
+                "tenant": tenant,
+                "database": database,
+                "collection": collection,
+            }
+        )
+        return
+
+    @trace_method(
+        "FastAPI.resolve_tenant_and_databases", OpenTelemetryGranularity.OPERATION
+    )
+    async def resolve_tenant_and_databases(
+        self,
+        request: Request,
+    ) -> UserIdentity:
+        if not self.authn_provider:
+            return UserIdentity(
+                user_id="", tenant=DEFAULT_TENANT, databases=[DEFAULT_DATABASE]
+            )
+
+        identity = self.authn_provider.authenticate_or_raise(dict(request.headers))
+        return cast(UserIdentity, await to_thread.run_sync(lambda: identity))
```

Comment:
Whoops, originally I had the call to `authenticate_or_raise` within the lambda so that we could make the network call without blocking. I think we should probably do this in case latency spikes.

Reply:
We can put it in the lambda too, thats likely "best" but as it was it was not needed yeah